api/dao/user_dao.js

```diff
@@ -97,41 +91,96 @@ module.exports = {
     },
 
     getUser: function(username, callback) {
-    // TODO figure out API to integrate with passport
-
-    // Look for tutors with the same username
-        Tutor.find({_id: username}, function (err, docs) {
+      // Look for tutors with the same username
+      Tutor.find({_id: username}, function (err, docs) {
+        if (err) {
+          console.error('Error checking if username is taken:', err);
+          callback(err);
+
+        } else if (docs.length === 1) {
+          // Found a tutor with the same username
+          callback(null, docs[0]);
+        } else if (docs.length > 1) {
+          console.error('Multiple tutors with username', username);
+
+        } else {
+
+          // Look for students with the same username
+          Student.find({_id: username}, function (err, docs) {
             if (err) {
-                console.error('Error checking if username is taken:', err);
-                callback(err);
+              console.error('Error checking if username is taken:', err);
+              callback(err);
 
             } else if (docs.length === 1) {
-                // Found a tutor with the same username
-                callback(null, docs[0]);
+              // Found a student with the same username
+              callback(null, docs[0]);
             } else if (docs.length > 1) {
-                console.error('Multiple tutors with username', username);
+              console.error('Multiple students with username', username);
 
             } else {
+              // No tutors or students with that username!
+              callback(null, null);
+            }
+          });
+        }
+      });
+    },
 
-                // Look for students with the same username
-                Student.find({_id: username}, function (err, docs) {
-                    if (err) {
-                        console.error('Error checking if username is taken:', err);
-                        callback(err);
+    /**
+     * TODO: Availability is not used currently
+     * TODO: No grade level checking for subject
+     */
+    getAllAvailableTutors: function(subject, availability, callback) {
```

Comment:
So i am a little confused, why do we have a function with functions in them?

Reply:
Accessing mongodb is async, we can't do it synchronously. You pass in a callback that gets called when the tutors are all retrieved from mongo. So you would do something like this:

```javascript
function main() {
dao.getAllAvailableTutors('CS', 'Tuesdays', printTutors);
}

function printTutors(err, tutors) {
  console.log(tutors);
}```

Reply:
The functions in the function are just boolean functions used for filtering. They're used by the ```tutors.filter``` lines near the end of the main function. You can have functions inside functions in javascript